crates/typst-ide/src/definition.rs

```diff
@@ -0,0 +1,240 @@
+            }
+        }
+    });
+    let src = src.or_else(|| {
+        let in_math = matches!(
+            leaf.parent_kind(),
+            Some(SyntaxKind::Equation)
+                | Some(SyntaxKind::Math)
+                | Some(SyntaxKind::MathFrac)
+                | Some(SyntaxKind::MathAttach)
+        );
+        let library = world.library();
+
+        let scope = if in_math { library.math.scope() } else { library.global.scope() };
+        for (item_name, value) in scope.iter() {
+            if item_name == &name {
+                return Some(Definition::item(
+                    name.clone(),
+                    Span::detached(),
+                    Some(value.clone()),
+                ));
+            }
+        }
+
+        None
+    });
+
+    if paths.is_empty() {
+        return src;
+    }
+
+    None
+}
+
+/// A definition of some item.
+#[derive(Debug, Clone)]
+pub struct Definition {
+    /// The name of the definition.
+    pub name: EcoString,
+    /// The kind of the definition.
+    pub kind: DefinitionKind,
+    /// A possible instance of the definition.
+    pub value: Option<Value>,
+    /// The source range of the entire definition.
+    pub range: Span,
+    /// The range of the name of the definition.
+    pub name_range: Span,
+}
+
+impl Definition {
+    fn with_kind(self, kind: DefinitionKind) -> Self {
+        Self { kind, ..self }
+    }
+
+    fn item(name: EcoString, span: Span, value: Option<Value>) -> Self {
+        let var_kind = value
+            .as_ref()
+            .and_then(|e| {
+                if matches!(e, Value::Func(..)) {
+                    Some(ItemDefinitionKind::Function)
+                } else {
+                    None
+                }
+            })
+            .unwrap_or(ItemDefinitionKind::Variable);
+
+        Self {
+            name,
+            kind: DefinitionKind::Item(var_kind),
+            value,
+            range: span,
+            name_range: span,
+        }
+    }
+
+    fn module(module: &Module) -> Self {
+        let file_id = match module.file_id() {
+            Some(file_id) => ModuleDefinitionKind::Path(file_id),
+            None => ModuleDefinitionKind::Builtin(module.clone()),
+        };
+
+        Definition {
+            name: EcoString::new(),
+            kind: DefinitionKind::Module(file_id),
+            value: Some(Value::Module(module.clone())),
+            range: Span::detached(),
+            name_range: Span::detached(),
+        }
+    }
+}
+
+/// The kind of a definition.
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub enum DefinitionKind {
+    /// `import "foo";`
+    ///         ^^^^^
+    Module(ModuleDefinitionKind),
+    /// `let foo;`
+    ///      ^^^
+    Item(ItemDefinitionKind),
+    /// `<foo>`
+    ///   ^^^
+    Label,
+}
+
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub enum ModuleDefinitionKind {
+    /// `import ("fo" + "o.typ")`
+    ///           ^^^^^^^^
+    Path(FileId),
```

Comment:
Since this is constructed from a `Module`, why are we just providing the FileId, instead of the full module?

Reply:
We may not be able to provide a `Module` if evaluating the module fails. It looks better if we resolve as a `FileId` with evaluation whenever possible. I did it in `tinymist-query` but I deleted the related code to reduce cost to the first PR.